sklearn/_fast_gradient_boosting/_binning.pyx

```diff
@@ -0,0 +1,58 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: nonecheck=False
+# cython: language_level=3
+
+# Author: Nicolas Hug
+
+cimport cython
+
+import numpy as np
+cimport numpy as np
+from cython.parallel import prange
+
+from .types cimport X_DTYPE_C, X_BINNED_DTYPE_C
+
+cpdef _map_to_bins(const X_DTYPE_C [:, :] data, list binning_thresholds,
+                   X_BINNED_DTYPE_C [::1, :] binned):
+    """Bin numerical values to discrete integer-coded levels.
+
+    Parameters
+    ----------
+    data : array-like, shape=(n_samples, n_features)
+        The numerical data to bin.
+    binning_thresholds : tuple of arrays
+        For each feature, stores the increasing numeric values that are
+        used to separate the bins.
+    binned : array-like, shape=(n_samples, n_features)
+        Output array, must be fortran aligned.
+    """
+    cdef:
+        int feature_idx
+
+    for feature_idx in range(data.shape[1]):
+        _map_num_col_to_bins(data[:, feature_idx],
+                             binning_thresholds[feature_idx],
+                             binned[:, feature_idx])
+
+
+cpdef void _map_num_col_to_bins(const X_DTYPE_C [:] data,
+                                const X_DTYPE_C [:] binning_thresholds,
+                                X_BINNED_DTYPE_C [:] binned):
+    """Binary search to the find the bin index for each value in data."""
+    cdef:
+        int i
+        int left
+        int right
+        int middle
+
+    for i in prange(data.shape[0], schedule='static', nogil=True):
```

Comment:
IIRC, if you have less then the number of bins distinct values, you put them each in their own bin. Wouldn't that make it possible to have an `O(1)` retrieval instead of `O(log(n))`? Not sure if/how easy we could gain that speed here though. Just a thought!

Reply:
Interesting!

I don't think this is possible since the unique values are simply used to computed the bins **thresholds**. But the `binned` array still needs to be integer-values, and the mapping from `data` (float values) to `binned` (int values) is never trivial.

Also I can see this king of reasoning break in a `fit(X_1).transform(X_2)` context: features that have less than 256 unique values in X_1 may have more unique values in X_2.